app/src/main/java/ai/elimu/appstore/ui/applications/ApplicationListActivity.java

```diff
@@ -65,13 +65,21 @@ protected void onStart() {
         RoomDb.databaseWriteExecutor.execute(() -> {
             List<Application> applications = applicationDao.loadAll();
             Log.d(getClass().getName(), "applications.size(): " + applications.size());
-            applicationListAdapter.setApplications(applications);
-
+            if(null!=applications) {
+                applicationListAdapter.setApplications(applications);
+            }
             List<ApplicationVersion> applicationVersions = applicationVersionDao.loadAll();
             Log.d(getClass().getName(), "applicationVersions.size(): " + applicationVersions.size());
             applicationListAdapter.setApplicationVersions(applicationVersions);
 
-            applicationListAdapter.notifyDataSetChanged();
+
+            recyclerView.post(new Runnable() {
+                @Override
+                public void run() {
+                    applicationListAdapter.notifyDataSetChanged();
```

Comment:
Why was this code moved into a thread?

Reply:
To make it work on UI thread where we are updating the data on to UI.

Reply:
setApplication synchronously adds a message to the UI message queue to render items. But the message will be handled later, not when you call setApplication.

If you call notifyItemChanged right away, then it will be called before actual rendering because the message generated by setApplication hasn't been dispatched yet.